Client generation with required sanitization

Sachindu-Nethmin opened the pull request:
## Purpose
To add spotify icon.png file and change config file.
## Examples

## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests
- [x] Updated the spec
- [ ] Checked native-image compatibility

Sachindu-Nethmin committed:
update icon.png

I added spotify icon.png file and change config file.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ballerina-platform/module-ballerinax-spotify?pullRequest=2) <br/>All committers have signed the CLA.

Sachindu-Nethmin committed:
Update README file with Overview

Enhance README.md with Overview

Sachindu-Nethmin committed:
Update README.md

aashikam commented:
hi @Sachindu-Nethmin,

This looks great. Could you try and follow this approach in opening pull requests for the different areas mentioned here in this [Ballerina hacktoberfest guide](https://ballerina.io/hacktoberfest/connector-contributor-guide/#step-4-contribution-workflow).

Sachindu-Nethmin committed:
Client generation with required sanitization

Sachindu-Nethmin reviewed (commented):
Client generation with required sanitization

Sachindu-Nethmin commented:
@aashikam can you please check it again please leave me a comment.

aashikam commented:
Hi @Sachindu-Nethmin,

LGTM 👍, Could you change the title for something more appropriate.

Sachindu-Nethmin commented:
@aashikam, your suggestion is greatly appreciated! I will make sure to update the title to better fit the content. Your feedback means a lot to me!

aashikam commented:
@Sachindu-Nethmin Looks good!, I will review it. In the meantime could you add a sanitization doc with the changes you made (if there are any) as mentioned here: https://ballerina.io/learn/create-your-first-connector-with-ballerina/#step-2-prepare-the-openapi-specification. 

You have to edit this: https://github.com/ballerina-platform/module-ballerinax-spotify/blob/main/docs/spec/sanitations.md

Here is an example of the doc: https://github.com/ballerina-platform/module-ballerinax-googleapis.gmail/blob/master/docs/spec/sanitation.md

aashikam commented on docs/spec/openapi.yaml:
```diff
@@ -0,0 +1,6930 @@
+        width:
+          type: integer
+          example: 300
+          nullable: t
```
Add a newline at the end of the spec please.

Sachindu-Nethmin committed:
Update openapi.yaml with new lines

aashikam commented on docs/spec/openapi.yaml:
```diff
@@ -0,0 +1,6930 @@
+              example: /9j/2wCEABoZGSccJz4lJT5CLy8vQkc9Ozs9R0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0cBHCcnMyYzPSYmPUc9Mj1HR0dEREdHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR//dAAQAAf/uAA5BZG9iZQBkwAAAAAH/wAARCAABAAEDACIAAREBAhEB/8QASwABAQAAAAAAAAAAAAAAAAAAAAYBAQAAAAAAAAAAAAAAAAAAAAAQAQAAAAAAAAAAAAAAAAAAAAARAQAAAAAAAAAAAAAAAAAAAAD/2gAMAwAAARECEQA/AJgAH//Z
+              format: byte
+              type: string
+              required: []
```
I believe you encountered this error here:
```
OpenAPI definition has errors: 
attribute paths.'/playlists/{playlist_id}/images'(put).requestBody.content.'image/jpeg'.schema.required is not of type `array`
```
and you have fixed it by changing `required: true` to `required: []`. This might not be the ideal solution. 
According to the open API spec:

**Understanding the Error**

The error message you encountered is:
```
OpenAPI definition has errors: 
attribute paths.'/playlists/{playlist_id}/images'(put).requestBody.content.'image/jpeg'.schema.required is not of type `array`
```

This indicates that the required attribute within the schema for the image/jpeg content type is incorrectly defined. In OpenAPI, the required attribute within a schema is expected to be an array when defining required properties of an object. However, in your case, the schema is defining a primitive type (`string`), not an object with properties.

**Correct Usage of `required`**
1. For Object Schemas:

When defining an object with properties, required should be an array listing the names of the properties that are mandatory.

```yaml
type: object
properties:
  name:
    type: string
  age:
    type: integer
required:
  - name
```

2. For Request Bodies:

To indicate that the entire request body is required, use the required attribute at the requestBody level, not within the schema.

```yaml
requestBody:
  required: true
  content:
    image/jpeg:
      schema:
        type: string
        format: byte
        description: Base64 encoded JPEG image data, maximum payload size is 256 KB.
```

3. For Primitive Schemas:

The required attribute should not be used within the schema when the schema defines a primitive type like string, integer, etc.

**How to Fix Your OpenAPI Definition**

Remove the required Field from the Schema:

```yaml
put:
  tags:
    - Playlists
  operationId: upload-custom-playlist-cover
  summary: |
    Add Custom Playlist Cover Image
  description: |
    Replace the image used to represent a specific playlist.
  parameters:
    - $ref: '#/components/parameters/PathPlaylistId'
  requestBody:
    required: true  # Ensure the request body itself is required
    content:
      image/jpeg:
        schema:
          type: string
          format: byte
          example: /9j/2wCEABoZGSccJz4lJT5CLy8vQkc9Ozs9R0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0cBHCcnMyYzPSYmPUc9Mj1HR0dEREdHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR//dAAQAAf/uAA5BZG9iZQBkwAAAAAH/wAARCAABAAEDACIAAREBAhEB/8QASwABAQAAAAAAAAAAAAAAAAAAAAYBAQAAAAAAAAAAAAAAAAAAAAAQAQAAAAAAAAAAAAAAAAAAAAARAQAAAAAAAAAAAAAAAAAAAAD/2gAMAwAAARECEQA/AJgAH//Z
          description: Base64 encoded JPEG image data, maximum payload size is 256 KB.
  responses:
    '202':
      description: Image uploaded
    '401':
      $ref: '#/components/responses/Unauthorized'
    '403':
      $ref: '#/components/responses/Forbidden'
    '429':
      $ref: '#/components/responses/TooManyRequests'
  security:
    - oauth_2_0:
        - ugc-image-upload
        - playlist-modify-public
        - playlist-modify-private
```

> Ensure required is Appropriately Placed: The required: true should be at the requestBody level to indicate that the request body is mandatory.

Hope this helps.

aashikam commented:
I believe you encountered this error here:
```
OpenAPI definition has errors: 
attribute paths.'/playlists/{playlist_id}/images'(put).requestBody.content.'image/jpeg'.schema.required is not of type `array`
```
and you have fixed it by changing `required: true` to `required: []`. This might not be the ideal solution. 
According to the open API spec:

**Understanding the Error**

The error message you encountered is:
```
OpenAPI definition has errors: 
attribute paths.'/playlists/{playlist_id}/images'(put).requestBody.content.'image/jpeg'.schema.required is not of type `array`
```

This indicates that the required attribute within the schema for the image/jpeg content type is incorrectly defined. In OpenAPI, the required attribute within a schema is expected to be an array when defining required properties of an object. However, in your case, the schema is defining a primitive type (`string`), not an object with properties.

**Correct Usage of `required`**
1. For Object Schemas:

When defining an object with properties, required should be an array listing the names of the properties that are mandatory.

```yaml
type: object
properties:
  name:
    type: string
  age:
    type: integer
required:
  - name
```

2. For Request Bodies:

To indicate that the entire request body is required, use the required attribute at the requestBody level, not within the schema.

```yaml
requestBody:
  required: true
  content:
    image/jpeg:
      schema:
        type: string
        format: byte
        description: Base64 encoded JPEG image data, maximum payload size is 256 KB.
```

3. For Primitive Schemas:

The required attribute should not be used within the schema when the schema defines a primitive type like string, integer, etc.

**How to Fix Your OpenAPI Definition**

Remove the required Field from the Schema:

```yaml
put:
  tags:
    - Playlists
  operationId: upload-custom-playlist-cover
  summary: |
    Add Custom Playlist Cover Image
  description: |
    Replace the image used to represent a specific playlist.
  parameters:
    - $ref: '#/components/parameters/PathPlaylistId'
  requestBody:
    required: true  # Ensure the request body itself is required
    content:
      image/jpeg:
        schema:
          type: string
          format: byte
          example: /9j/2wCEABoZGSccJz4lJT5CLy8vQkc9Ozs9R0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0cBHCcnMyYzPSYmPUc9Mj1HR0dEREdHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR0dHR//dAAQAAf/uAA5BZG9iZQBkwAAAAAH/wAARCAABAAEDACIAAREBAhEB/8QASwABAQAAAAAAAAAAAAAAAAAAAAYBAQAAAAAAAAAAAAAAAAAAAAAQAQAAAAAAAAAAAAAAAAAAAAARAQAAAAAAAAAAAAAAAAAAAAD/2gAMAwAAARECEQA/AJgAH//Z
          description: Base64 encoded JPEG image data, maximum payload size is 256 KB.
  responses:
    '202':
      description: Image uploaded
    '401':
      $ref: '#/components/responses/Unauthorized'
    '403':
      $ref: '#/components/responses/Forbidden'
    '429':
      $ref: '#/components/responses/TooManyRequests'
  security:
    - oauth_2_0:
        - ugc-image-upload
        - playlist-modify-public
        - playlist-modify-private
```

> Ensure required is Appropriately Placed: The required: true should be at the requestBody level to indicate that the request body is mandatory.

Hope this helps.

Sachindu-Nethmin committed:
Update openapi.yaml

Remove the required Field from the Schema

Sachindu-Nethmin committed:
Update sanitations.md

add a sanitization doc with the changes I made

aashikam commented:
Hi @Sachindu-Nethmin,

You seem to have attempted to remove the incorrect resource (`/playlists/{playlist_id}/images (put)`) entirely from the spec without editing it. Shall we edit the spec with the correct format and update the PR. Also the segment you have removed  in this commit doesn't seem to be the incorrect one anyway. https://github.com/ballerina-platform/module-ballerinax-spotify/pull/2/commits/0f96f2e3e451042ab5ac6d495bb15d37c357f858

The code changes are good to go. Once the spec is updated I will merge this PR. 

Feel free to reach out for any doubts 🙌 I am here to help.

aashikam reviewed (dismissed):
LGTM

Sachindu-Nethmin commented:
@aashikam I want to continue contributing to open-source projects even after completing my current work. Even though I'm not sure if it's possible or not,  I would appreciate your assistance. This opportunity could be life-changing for me.

NipunaRanasinghe commented:
> @aashikam I want to continue contributing to open-source projects even after completing my current work. Even though I'm not sure if it's possible or not, I would appreciate your assistance. This opportunity could be life-changing for me.

@Sachindu-Nethmin It's wonderful to hear that you want to continue contributing to our open-source projects! 

Ballerina is fully open-source, and we always welcome contributions from the community. Here are some ways you can keep contributing:

- We're planning to extend this connector event beyond Hacktoberfest , so once you complete your current work, you’ll have more opportunities to contribute to more connectors (with rewards included)
- You can keep an eye on our "Good First Issues" that might interest you:
    - [Ballerina Lang Good First Issues](https://github.com/ballerina-platform/ballerina-lang/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22)
    - [Ballerina Library Good First Issues](https://github.com/ballerina-platform/ballerina-library/issues?q=is%3Aopen+is%3Aissue+label%3A%22Good+first+issue%22)
- If you happen to be an undergraduate and are interested in representing Ballerina in your university/institute, you might consider joining the [Ballerina Ambassador Program](https://ballerina.io/community/ambassadors/).

Excited to see your contributions grow! 🚀 Let me know if you need any help.

NipunaRanasinghe commented on ballerina/client.bal:
```diff
@@ -1,15 +1,993 @@
+    #
+    # + headers - Headers to be sent with the request 
+    # + return - A markets object with an array of country codes 
+    resource isolated function get markets(map<string|string[]> headers = {}) returns inline_response_200_16|error {
```
@Sachindu-Nethmin we don't prefer having these `inline_response_*` record types as the response type in connector APIs. This happens when the particular reponse type is defined as an inline schema in the OpenAPI contract. 

So in order to generated the response type with a proper name, you can update the OpenAPI spec file by moving the particular inline schema to the `components` section. 
Also Please make sure to mention record these improvements in the `sanitations.md` file.

NipunaRanasinghe commented on ballerina/client.bal:
```diff
@@ -1,15 +1,993 @@
+    # + headers - Headers to be sent with the request 
+    # + queries - Queries to be sent with the request 
+    # + return - Show saved 
+    resource isolated function put me/shows(map<string|string[]> headers = {}, *SaveShowsUserQueries queries) returns http:Response|error {
```
it seems some APIs are returning `http:Response` instead of proper record types. @lnash94 @TharmiganK need to verify if this is the expected behaviour or, a potential issue in the OpenAPI client generator.
```suggestion
    resource isolated function put me/shows(map<string|string[]> headers = {}, *SaveShowsUserQueries queries) returns http:Response|error {
```

lnash94 replied:
In this case, the given operation has a success status code without a specific body, and in the HTTP client, we cannot directly bind the status code as `http:Ok`. The `http:Response` is used because the operation has multiple responses.

If the user is interested in binding the status code, they need to add the `—status-code-binding` flag when generating the code.

Since currently we only consider the success payload, We can check the feasibility of improving the tool to generate `error?` for multiple responses. In the meantime, until we provide this improvement, we can modify the OAS by removing specific error responses.

[1] OAS sample which success response has no content type; Here success response is 200 and it only has a description. Remaining of the status codes are error status code. 
```yml
put:
      tags:
        - Shows
        - Library
      operationId: save-shows-user
      summary: |
        Save Shows for Current User
      description: |
        Save one or more shows to current Spotify user's library.
      parameters:
        - $ref: '#/components/parameters/QueryShowIds'
      responses:
        '200':
          description: Show saved
        '401':
          $ref: '#/components/responses/Unauthorized'
        '403':
          $ref: '#/components/responses/Forbidden'
        '429':
          $ref: '#/components/responses/TooManyRequests'
      security:
        - oauth_2_0:
            - user-library-modify
```
[2] Modified OAS: Simply removed the remaining status code 401, 403, 429 since we bind and process data with the success payloads.

```yml
put:
      tags:
        - Shows
        - Library
      operationId: save-shows-user
      summary: |
        Save Shows for Current User
      description: |
        Save one or more shows to current Spotify user's library.
      parameters:
        - $ref: '#/components/parameters/QueryShowIds'
      responses:
        '200':
          description: Show saved
      security:
        - oauth_2_0:
            - user-library-modify
```

NipunaRanasinghe commented on docs/spec/sanitations.md:
```diff
@@ -1,24 +1,24 @@
-1. 
-2. 
-3. 
+1. Remove resource paths,
+    */playlists/{playlist_id}/images'(put) - schema for the image/jpeg content type is incorrectly defined
```
@Sachindu-Nethmin can you elaborate more on this?

NipunaRanasinghe replied:
@Sachindu-Nethmin could you please explain the reason to remove this resource path? Just checking to see if we have any alternative instead of just removing this API support from the connector.

Sachindu-Nethmin replied:
@NipunaRanasinghe 
  In '/playlists/{playlist_id}/images'
    Remove the required Field from the Schema because the required attribute within the schema for the image/jpeg content type is incorrectly defined. I think I did it in an alternative way now.

NipunaRanasinghe commented on docs/spec/sanitations.md:
````diff
@@ -1,24 +1,24 @@
 
 ```bash
-# TODO: Add OpenAPI CLI command used to generate the client
+bal openapi -i docs/spec/openapi.yaml --mode client  --license docs/license.txt -o ballerina/modules/oas
````
```suggestion
bal openapi -i docs/spec/openapi.yaml --mode client --license docs/license.txt -o ballerina
```

Sachindu-Nethmin committed:
Update sanitations.md

Update openAPI CLI command used to generate the client

Sachindu-Nethmin committed:
Update sanitations.md

Move inline response schemas to the components section

Sachindu-Nethmin committed:
Remove duplicated responces

Sachindu-Nethmin committed:
Regenerated client

Sachindu-Nethmin committed:
Remove inline response schemas

aashikam commented:
Hi @Sachindu-Nethmin,

I feel like you have misunderstood @NipunaRanasinghe's suggestion:
> @Sachindu-Nethmin we don't prefer having these `inline_response_*` record types as the response type in connector APIs. This happens when the particular reponse type is defined as an inline schema in the OpenAPI contract.
> 
> So in order to generated the response type with a proper name, you can update the OpenAPI spec file by moving the particular inline schema to the `components` section. Also Please make sure to mention record these improvements in the `sanitations.md` file.

**Understanding better:**

```ballerina
resource isolated function get markets(map<string|string[]> headers = {}) returns inline_response_200_16|error {
```
If you check this particular resource function for example you can see it has this `returns inline_response_200_16|error` part at the end. What Nipuna is suggesting is that we don't usually use this `inline_response_200_16` as the return type. This `inline_response_200_16` is auto-generated by our OpenAPI tool as it is defined as an inline schema in the OpenAPI specification. 

**Open API spec segment where you need to edit:**
```yaml
  /markets:
    get:
      tags:
        - Markets
      operationId: get-available-markets
      summary: |
        Get Available Markets
      description: |
        Get the list of markets where Spotify is available.
      responses:
        '200':
          description: A markets object with an array of country codes
          content:
            application/json:
              schema:
                type: object
                properties:
                  markets:
                    type: array
                    items:
                      type: string
                    example: [ 'CA', 'BR', 'IT' ]
        '401':
          $ref: '#/components/responses/Unauthorized'
        '403':
          $ref: '#/components/responses/Forbidden'
        '429':
          $ref: '#/components/responses/TooManyRequests'
      security:
        - oauth_2_0: [ ]
``` 
**Corrected Open API spec based on the suggestion:**
```yaml
// added in components.schemas section
components:
  schemas:
    MarketsResponse:
      type: object
      properties:
        markets:
          type: array
          items:
            type: string
      example:
        markets: [ 'CA', 'BR', 'IT' ]
```
```yaml
paths:
  /markets:
    get:
      tags:
        - Markets
      operationId: get-available-markets
      summary: |
        Get Available Markets
      description: |
        Get the list of markets where Spotify is available.
      responses:
        '200':
          description: A markets object with an array of country codes
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/MarketsResponse'
        '401':
          $ref: '#/components/responses/Unauthorized'
        '403':
          $ref: '#/components/responses/Forbidden'
        '429':
          $ref: '#/components/responses/TooManyRequests'
      security:
        - oauth_2_0: [ ]

```

This is an example fix for one API, this needs to be done in all the places where something equivalent to `inline_response_200_16` is observed. After making the necessary changes to the specification, regenerating the client will fix the necessary code quality requirement. 

Feel free to reach out. 🙌

Sachindu-Nethmin committed:
revert changes on openapi.yaml

Sachindu-Nethmin committed:
Update openapi.yaml and regenerate client

    1.MarketsResponse
    2.PlaylistResponse
    3.ArtistObjects
    4.ManyTrackObject
    5.ManyAudioFeaturesObject
    6.PagedCategoriesObject
    7.PagedAlbumsObject
    8.CursorPagedArtistsObject
    9.ManyDevicesObject
    10.PagingArtistOrTrackObject
    11.ManyGenresObject
    12.SearchItemsObject
    13.ManySimplifiedShowsObject
    14.ManyAlbumsObject
    15.ManyAudiobooksObject
    16.ManyChaptersObject
    17.ManyEpisodesObject
    18.PlaylistSnapshotIdObject

Sachindu-Nethmin commented:
@aashikam, I would appreciate your feedback on whether I’m on the right path. This guidance will empower me to move forward successfully with the task.

NipunaRanasinghe commented on docs/spec/sanitations.md:
```diff
@@ -12,16 +12,10 @@ These changes are done in order to improve the overall usability, and as workaro
-        '403':
-          $ref: '#/components/responses/Forbidden'
-        '429':
-          $ref: '#/components/responses/TooManyRequests'
```
@Sachindu-Nethmin we need to keep the sanitation number 3 (Remove bad responses to increase readability) so don't revert it. 

We only need to do what @aashikam mentioned in https://github.com/ballerina-platform/module-ballerinax-spotify/pull/2#issuecomment-2435443820, instead of removing header and query parameters

aashikam commented:
Hi @Sachindu-Nethmin,

The changes you have done in 4a3b60dbd1aee4c6fd99f88582f0805358268302 seems to be correct. Please repeat the same in places where applicable.

Sachindu-Nethmin commented:
Could you explain how to identify an inline schema within the OpenAPI contract?

aashikam commented:
Hi @Sachindu-Nethmin,

The correct way to identify the inline responses in the Open API specification is to go through the specification, identify where it has defined the responses without a specific schema type. 

**What you should look for:**
```yaml
responses:
        '200':
          description: A markets object with an array of country codes
          content:
            application/json:
              schema:
                type: object
                properties:
                  markets:
                    type: array
                    items:
                      type: string
                    example: [ 'CA', 'BR', 'IT' ]
```

**Which should be corrected as:**

```yaml
      responses:
        '200':
          description: A markets object with an array of country codes
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/MarketsResponse'
```
`MarketsResponse` should be defined  under components.schemas.

But there is an easier way. Since you have already generated the client, you can search for `inline_response` in the `client.bal` code and edit the corresponding Open API definition. 😁
<img width="931" alt="Screenshot 2024-10-25 at 11 39 17" src="https://github.com/user-attachments/assets/cf8eb052-4783-46bf-89f7-38bf9a249dfa">

Sachindu-Nethmin commented:
I already did that but can not find which lines should move in .yaml file.

Sachindu-Nethmin commented:
@aashikam Please review it once more.

aashikam reviewed (dismissed):
LGTM

Shall we squash the commits please. 

<img width="1358" alt="Screenshot 2024-10-28 at 22 00 25" src="https://github.com/user-attachments/assets/6ccaeac3-7720-44c7-b494-6b81b9b3b568">

As a best practice, keeping commits to a minimum helps maintain a clean and understandable project history. Squashing commits combines multiple smaller commits into a single, more meaningful one.

**How to Squash Commits:**

1. Identify the Number of Commits to Squash:

Determine how many commits you want to combine. Let's say you want to squash the last 3 commits.

2. Start an Interactive Rebase:

Run the command:

```sh
git rebase -i HEAD~3
```

This opens an interactive editor with a list of the last 3 commits.

3. Modify the Commit Commands:

In the editor, you'll see each commit listed with the word pick before it.
Change `pick` to `squash` (or s) for the commits you want to squash into the one above it. For example:

```sh
pick abc1234 Commit message 1
squash def5678 Commit message 2
squash ghi9012 Commit message 3
```

4. Save and Close the Editor:

This initiates the squashing process.

5. Edit the Commit Message:

Another editor window will open, allowing you to combine and edit the commit messages as needed.
Write a concise and descriptive commit message for the combined commit.

6. Finalize the Rebase:

Save and close the editor to complete the rebase process.

7. Force Push the Changes (if necessary):

If you've already pushed the original commits to a remote repository, you'll need to force push to update the remote history:

```sh
git push --force
```

By squashing commits, you create a cleaner commit history that makes it easier for others to review changes and understand the project's progression.

> Easier hack: You can achieve the same result using git reset --soft to uncommit the desired commits while keeping their changes staged. For example, to squash the last three commits into one, run git reset --soft HEAD~3, and then commit the changes anew with git commit -m "Combined commit message". This effectively combines the changes from those commits into a single new commit.

aashikam commented:
Need @NipunaRanasinghe approval, since he requested changes.

aashikam commented:
Hey @Sachindu-Nethmin!
Hope everything’s going well so far! Just a quick update on the project deadlines: it's set for one month from the day you were officially assigned. But if you encounter any challenges or unforeseen issues, please feel free to reach out, and we can consider adjusting the deadline accordingly.

If you have any questions or need help with anything, please feel free to reach out. Keep up the awesome work! 🚀

NipunaRanasinghe commented:
> Need @NipunaRanasinghe approval, since he requested changes.

@aashikam once he addresses https://github.com/ballerina-platform/module-ballerinax-spotify/pull/2#discussion_r1821256545 we are good to merge this. @Sachindu-Nethmin FYA

Sachindu-Nethmin committed:
Update openapi.yaml

Remove the required Field from the Schema because required attribute within the schema for the image/jpeg content type is incorrectly defined.

Sachindu-Nethmin committed:
Regenerate client

Sachindu-Nethmin committed:
Update sanitations.md

Sachindu-Nethmin commented:
@aashikam, could you please guide me on the steps I need to take?

aashikam reviewed (approved):
LGTM, Verified the spec against the original/official and the generated code changes. @NipunaRanasinghe Need your approval since you have requested changes.

NipunaRanasinghe reviewed (approved):
LGTM

aashikam commented:
@Sachindu-Nethmin Well done! 🚀

You can proceed with the next steps. Refer to this guide for more detailed explanation. [Ballerina hacktoberfest guide](https://ballerina.io/hacktoberfest/connector-contributor-guide/#step-4-contribution-workflow).
